framework/db/src/com/cloud/utils/db/Transaction.java

```diff
@@ -59,4 +59,30 @@ public T doInTransaction(TransactionStatus status) throws RuntimeException {
         });
     }
 
+    @SuppressWarnings("deprecation")
+    public static <T, E extends Throwable> T execute(final short databaseId, TransactionCallbackWithException<T, E> callback) throws E {
```

Comment:
``public static <T, E extends Throwable> T execute(TransactionCallbackWithException<T, E> callback) throws E`` should be implemented in terms of this method.  There is a tremendous amount of code duplication between the two methods.